src/apphosting/secrets/dialogs.ts

```diff
@@ -0,0 +1,148 @@
+  ];
+}
+
+export function toMetadata(projectNumber: string, backends: apphosting.Backend[]): Metadata[] {
+  const metadata: Metadata[] = [];
+  for (const backend of backends) {
+    const [, , , location, , id] = backend.name.split("/");
+    metadata.push({ location, id, accounts: serviceAccountsForBackend(projectNumber, backend) });
+  }
+  return metadata.sort((left, right) => {
+    const cmplocation = left.location.localeCompare(right.location);
+    if (cmplocation) {
+      return cmplocation;
+    }
+    return left.id.localeCompare(right.id);
+  });
+}
+
+export function tableForBackends(metadata: Metadata[]): [headers: string[], rows: string[][]] {
+  const headers = [
+    "location",
+    "backend",
+    metadata.findIndex((val) => val.accounts.length > 1) === -1
+      ? "service account"
+      : "service accounts",
+  ];
+  const rows = metadata.map((m) => [m.location, m.id, m.accounts.join(", ")]);
+  return [headers, rows];
+}
+
+export const WARN_NO_BACKENDS =
+  "To use this secret, your backend's service account must have secret accessor permission. " +
+  "It does not look like you have a backend yet. After creating a backend, grant access with " +
+  clc.bold("firebase apphosting:secrets:grantAccess");
+
+export const GRANT_ACCESS_IN_FUTURE = `To grant access in the future, run ${clc.bold("firebase apphosting:secrets:grantaccess")}`;
+
+export async function selectBackendServiceAccounts(
+  projectNumber: string,
+  projectId: string,
+  options: any,
+): Promise<string[]> {
+  const listBackends = await apphosting.listBackends(projectId, "-");
+
+  if (listBackends.unreachable.length) {
+    utils.logLabeledWarning(
+      "apphosting",
+      `Could not reach location(s) ${listBackends.unreachable.join(", ")}. You may need to run ` +
+        `${clc.bold("firebase apphosting:secrets:grantAccess")} at a later time if you have backends in these locations`,
+    );
+  }
+
+  if (!listBackends.backends.length) {
+    utils.logLabeledWarning("apphosting", WARN_NO_BACKENDS);
+    return [];
+  }
+
+  if (listBackends.backends.length === 1) {
+    const grant = await prompt.confirm({
+      nonInteractive: options.nonInteractive,
+      default: true,
+      message:
+        "To use this secret, your backend's service account must have secret accessor permission. Would you like to grant it now?",
+    });
+    if (grant) {
+      return serviceAccountsForBackend(projectNumber, listBackends.backends[0]);
+    }
+    utils.logLabeledBullet("apphosting", GRANT_ACCESS_IN_FUTURE);
+    return [];
+  }
+
+  const metadata: Metadata[] = toMetadata(projectNumber, listBackends.backends);
+
+  // Use JSON.stringify because deep copmarison is annoying in JS. Because the order of the service account list should be deterinistic,
+  // this shouldn't need a sort command.
+  const test = JSON.stringify(metadata[0].accounts);
+  const allSharedAccounts = metadata.every((val) => JSON.stringify(val.accounts) === test);
```

Comment:
Alternatively, you can do:

new Set(metadata.map(val => JSON.stringify(val.accounts)).size === 1

Reply:
Oh, forgot to mention here that `every` short circuits but `map` does not. It's a trivial optimization, but seems correct to use in this case.